File: framework/src/play-jdbc/src/main/scala/play/api/db/Databases.scala
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
 */
package play.api.db

import java.sql.{ Connection, Driver, DriverManager }
import java.util.concurrent.{ BlockingQueue, LinkedBlockingQueue, TimeUnit }
import javax.sql.DataSource

import com.typesafe.config.Config
import play.api.{ Configuration, Environment }
import play.core.Execution
import play.utils.{ ProxyDriver, Reflect }

import scala.annotation.tailrec
import scala.concurrent.{ ExecutionContext, Future, Promise }
import scala.util.control.{ ControlThrowable, NonFatal }
import scala.util.{ Failure, Success, Try }

/**
 * Creation helpers for manually instantiating databases.
 */
object Databases {

  /**
   * Create a pooled database named "default" with the given driver and url.
   *
   * @param driver the database driver class
   * @param url the database url
   * @param name the database name
   * @param config a map of extra database configuration
   * @return a configured database
   */
  def apply(driver: String, url: String, name: String = "default", config: Map[String, _ <: Any] = Map.empty): Database = {
    val dbConfig = Configuration.reference.get[Configuration]("play.db.prototype") ++
      Configuration.from(Map("driver" -> driver, "url" -> url) ++ config)
    new PooledDatabase(name, dbConfig)
  }

  /**
   * Create an in-memory H2 database.
   *
   * @param name the database name (defaults to "default")
   * @param urlOptions a map of extra url options
   * @param config a map of extra database configuration
   * @return a configured in-memory h2 database
   */
  def inMemory(name: String = "default", urlOptions: Map[String, String] = Map.empty, config: Map[String, _ <: Any] = Map.empty): Database = {
    val driver = "org.h2.Driver"
    val urlExtra = urlOptions.map { case (k, v) => k + "=" + v }.mkString(";", ";", "")
    val url = "jdbc:h2:mem:" + name + urlExtra
    Databases(driver, url, name, config)
  }

  /**
   * Run the given block with a database, cleaning up afterwards.
   *
   * @param driver the database driver class
   * @param url the database url
   * @param name the database name
   * @param config a map of extra database configuration
   * @param block The block of code to run
   * @return The result of the block
   */
  def withDatabase[T](driver: String, url: String, name: String = "default",
    config: Map[String, _ <: Any] = Map.empty)(block: Database => T): T = {
    val database = Databases(driver, url, name, config)
    try {
      block(database)
    } finally {
      database.shutdown()
    }
  }

  /**
   * Run the given block with an in-memory h2 database, cleaning up afterwards.
   *
   * @param name the database name (defaults to "default")
   * @param urlOptions a map of extra url options
   * @param config a map of extra database configuration
   * @param block The block of code to run
   * @return The result of the block
   */
  def withInMemory[T](name: String = "default", urlOptions: Map[String, String] = Map.empty,
    config: Map[String, _ <: Any] = Map.empty)(block: Database => T): T = {
    val database = inMemory(name, urlOptions, config)
    try {
      block(database)
    } finally {
      database.shutdown()
    }
  }
}

/**
 * Default implementation of the database API.
 * Provides driver registration and connection methods.
 */
abstract class DefaultDatabase(val name: String, configuration: Config, environment: Environment) extends Database {

  private val config = Configuration(configuration)
  val databaseConfig = DatabaseConfig.fromConfig(config, environment)

  // abstract methods to be implemented

  protected def createAsyncDataSource(): AsyncDataSource

  // driver registration

  lazy val driver: Option[Driver] = {
    databaseConfig.driver.map { driverClassName =>
      try {
        val proxyDriver = new ProxyDriver(Reflect.createInstance[Driver](driverClassName, environment.classLoader))
        DriverManager.registerDriver(proxyDriver)
        proxyDriver
      } catch {
        case NonFatal(e) => throw config.reportError("driver", s"Driver not found: [$driverClassName}]", Some(e))
      }
    }
  }

  // lazy data source creation

  override lazy val dataSource: AsyncDataSource = {
    driver // trigger driver registration
    createAsyncDataSource()
  }

  lazy val url: String = {
    val connection = dataSource.getConnection
    try {
      connection.getMetaData.getURL
    } finally {
      connection.close()
    }
  }

  // c methods

  override def getConnectionAsync(): Future[Connection] = {
    getConnectionAsync(autocommit = true)
  }

  override def getConnection(): Connection = {
    getConnection(autocommit = true)
  }

  override def getConnectionAsync(autocommit: Boolean): Future[Connection] = {
    dataSource.getConnectionAsync.map { connection =>
      connection.setAutoCommit(autocommit)
      connection
    }(Execution.trampoline)
  }

  override def getConnection(autocommit: Boolean): Connection = {
    val connection = dataSource.getConnection
    connection.setAutoCommit(autocommit)
    connection
  }

  override def withConnectionAsync[A](block: Connection => Future[A])(implicit ec: ExecutionContext): Future[A] = {
    withConnectionAsync(autocommit = true)(block)
  }

  override def withConnection[A](block: Connection => A): A = {
    withConnection(autocommit = true)(block)
  }

  override def withConnectionAsync[A](autocommit: Boolean)(block: Connection => Future[A])(implicit ec: ExecutionContext): Future[A] = {
    getConnectionAsync(autocommit).flatMap { connection =>
      block(connection).transform { result: Try[A] =>
        connection.close()
        result
      }(Execution.trampoline)
    }
  }

  override def withConnection[A](autocommit: Boolean)(block: Connection => A): A = {
    val connection = getConnection(autocommit)
    try {
      block(connection)
    } finally {
      connection.close()
    }
  }

  override def withTransactionAsync[A](block: Connection => Future[A])(implicit ec: ExecutionContext): Future[A] = {
    getConnectionAsync(autocommit = false).flatMap { connection =>
      block(connection).transform {
        case f @ Failure(_: ControlThrowable) =>
          connection.commit()
          f
        case f @ Failure(_) =>
          connection.rollback()
          f
        case s @ Success(_) =>
          connection.commit()
          s
      }(Execution.trampoline)
    }
  }

  override def withTransaction[A](block: Connection => A): A = {
    withConnection(autocommit = false) { connection =>
      try {
        val r = block(connection)
        connection.commit()
        r
      } catch {
        case e: ControlThrowable =>
          connection.commit()
          throw e
        case e: Throwable =>
          connection.rollback()
          throw e
      }
    }
  }

  override def shutdown(): Unit = {
    dataSource.close()
    driver.foreach(DriverManager.deregisterDriver)
  }

}

/**
 * Default implementation of the database API using a c pool.
 */
class PooledDatabase(name: String, configuration: Config, environment: Environment, pool: AsyncConnectionPool)
    extends DefaultDatabase(name, configuration, environment) {

  def this(name: String, configuration: Configuration) = this(name, configuration.underlying, Environment.simple(), new HikariCPConnectionPool(Environment.simple()))

  override protected def createAsyncDataSource(): AsyncDataSource = {
    pool.createAsync(name, databaseConfig, configuration)
  }

}

File: framework/src/play-java-jdbc/src/main/java/play/db/Databases.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package play.db;

import java.util.Map;

import com.google.common.collect.ImmutableMap;

/**
 * Creation helpers for manually instantiating databases.
 */
public final class Databases {
// Databases is a final class and not an interface because an interface cannot be declared final. Also, that should
// clarify why the class' constructor is private: we really don't want this class to be either instantiated or subclassed.
	private Databases() {}
    // ----------------
    // Creation helpers
    // ----------------

    /**
     * Create a pooled database with the given configuration.
     *
     * @param name the database name
     * @param driver the database driver class
     * @param url the database url
     * @param config a map of extra database configuration
     * @return a configured database
     */
    public static Database createFrom(String name, String driver, String url, Map<String, ? extends Object> config) {
        ImmutableMap.Builder<String, Object> dbConfig = new ImmutableMap.Builder<String, Object>();
        dbConfig.put("driver", driver);
        dbConfig.put("url", url);
        dbConfig.putAll(config);
        return new DefaultDatabase(name, dbConfig.build());
    }

    /**
     * Create a pooled database with the given configuration.
     *
     * @param name the database name
     * @param driver the database driver class
     * @param url the database url
     * @return a configured database
     */
    public static Database createFrom(String name, String driver, String url) {
        return createFrom(name, driver, url, ImmutableMap.<String, Object>of());
    }

    /**
     * Create a pooled database named "default" with the given configuration.
     *
     * @param driver the database driver class
     * @param url the database url
     * @param config a map of extra database configuration
     * @return a configured database
     */
    public static Database createFrom(String driver, String url, Map<String, ? extends Object> config) {
        return createFrom("default", driver, url, config);
    }

    /**
     * Create a pooled database named "default" with the given driver and url.
     *
     * @param driver the database driver class
     * @param url the database url
     * @return a configured database
     */
    public static Database createFrom(String driver, String url) {
        return createFrom("default", driver, url, ImmutableMap.<String, Object>of());
    }

    /**
     * Create an in-memory H2 database.
     *
     * @param name the database name
     * @param url the database url
     * @param config a map of extra database configuration
     * @return a configured in-memory h2 database
     */
    public static Database inMemory(String name, String url, Map<String, ? extends Object> config) {
        return createFrom(name, "org.h2.Driver", url, config);
    }

    /**
     * Create an in-memory H2 database.
     *
     * @param name the database name
     * @param urlOptions a map of extra url options
     * @param config a map of extra database configuration
     * @return a configured in-memory h2 database
     */
    public static Database inMemory(String name, Map<String, String> urlOptions, Map<String, ? extends Object> config) {
        String urlExtra = "";
        for (Map.Entry<String, String> option : urlOptions.entrySet()) {
            urlExtra += ";" + option.getKey() + "=" + option.getValue();
        }
        String url = "jdbc:h2:mem:" + name + urlExtra;
        return inMemory(name, url, config);
    }

    /**
     * Create an in-memory H2 database.
     *
     * @param name the database name
     * @param config a map of extra database configuration
     * @return a configured in-memory h2 database
     */
    public static Database inMemory(String name, Map<String, ? extends Object> config) {
        return inMemory(name, "jdbc:h2:mem:" + name, config);
    }

    /**
     * Create an in-memory H2 database.
     *
     * @param name the database name
     * @return a configured in-memory h2 database
     */
    public static Database inMemory(String name) {
        return inMemory(name, ImmutableMap.<String, Object>of());
    }

    /**
     * Create an in-memory H2 database with name "default".
     *
     * @param config a map of extra database configuration
     * @return a configured in-memory h2 database
     */
    public static Database inMemory(Map<String, ? extends Object> config) {
        return inMemory("default", config);
    }

    /**
     * Create an in-memory H2 database with name "default".
     *
     * @return a configured in-memory h2 database
     */
    public static Database inMemory() {
        return inMemory("default");
    }

    /**
     * Create an in-memory H2 database with name "default" and with
     * extra configuration provided by the given entries.
     *
     * @param k1 an H2 configuration key.
     * @param v1 configuration value corresponding to `k1`
     * @return a configured in-memory H2 database
     */
    public static Database inMemoryWith(String k1, Object v1) {
        return inMemory(ImmutableMap.of(k1, v1));
    }

    /**
     * Create an in-memory H2 database with name "default" and with
     * extra configuration provided by the given entries.
     *
     * @param k1 an H2 configuration key
     * @param v1 H2 configuration value corresponding to `k1`
     * @param k2 a second H2 configuration key
     * @param v2 a configuration value corresponding to `k2`
     * @return a configured in-memory H2 database
     */
    public static Database inMemoryWith(String k1, Object v1, String k2, Object v2) {
        return inMemory(ImmutableMap.of(k1, v1, k2, v2));
    }

    /**
     * Create an in-memory H2 database with name "default" and with
     * extra configuration provided by the given entries.
     *
     * @param k1 an H2 configuration key
     * @param v1 H2 configuration value corresponding to `k1`
     * @param k2 a second H2 configuration key
     * @param v2 a configuration value corresponding to `k2`
     * @param k3 a third H2 configuration key
     * @param v3 a configuration value corresponding to `k3`
     * @return a configured in-memory H2 database
     */
    public static Database inMemoryWith(String k1, Object v1, String k2, Object v2, String k3, Object v3) {
        return inMemory(ImmutableMap.of(k1, v1, k2, v2, k3, v3));
    }
}
